packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js

```diff
@@ -179,6 +180,111 @@ const Conversation = WebexPlugin.extend({
       });
   },
 
+  /**
+   * generate a deterministic HMAC for a reaction
+   * @param {Object} displayName displayName of reaction we are sending
+   * @param {Object} parent parent activity of reaction
+   * @returns {Promise<HMAC>}
+   */
+  createReactionHmac(displayName, parent) {
+    // not using webex.internal.encryption.getKey() because the JWK it returns does not have a 'k'
+    // property. we need jwk.k to correctly generate the HMAC
+    return this.webex.internal.encryption.unboundedStorage.get(parent.encryptionKeyUrl)
+      .then((keyString) => {
+        const key = JSON.parse(keyString);
+        // when we stringify this object, keys must be in this order to generate same HMAC as
+        // desktop clients
+        const formatjwk = {k: key.jwk.k, kid: key.jwk.kid, kty: key.jwk.kty};
+
+        const source = `${JSON.stringify(formatjwk)}${parent.id}${displayName}`;
+
+        const hmac = crypto
+          .createHmac('sha256', parent.id)
+          .update(source)
+          .digest('hex');
+
+        return Promise.resolve(hmac);
+      });
+  },
+
+  /**
+   * send add or delete reaction to convo service
+   * @param {Object} conversation
+   * @param {Object} reactionPayload
+   * @returns {Promise<Activity>}
+   */
+  sendReaction(conversation, reactionPayload) {
+    const url = this.getConvoUrl(conversation);
+    const convoWithUrl = Object.assign({}, conversation, {url});
+
+    if (!isObject(reactionPayload)) {
+      return Promise.reject(new Error('`object` must be an object'));
+    }
+
+    return this.prepare(reactionPayload, {
+      target: this.prepareConversation(convoWithUrl),
+      object: pick(reactionPayload, 'id', 'url', 'objectType')
+    })
+      .then((act) => this.submit(act));
+  },
+
+  /**
+   * delete a reaction
+   * @param {Object} conversation
+   * @param {Object} reactionId
+   * @returns {Promise<Activity>}
+   */
+  deleteReaction(conversation, reactionId) {
+    const deleteReactionPayload = {
+      actor: {objectType: 'person', id: this.webex.internal.device.userId},
+      object: {
+        id: reactionId,
+        objectType: 'activity'
+      },
+      objectType: 'activity',
+      target: {
+        id: conversation.id,
+        objectType: 'conversation'
+      },
+      verb: 'delete'
+    };
+
+    return this.sendReaction(conversation, deleteReactionPayload);
+  },
+
+  /**
+   * create a reaction
+   * @param {Object} conversation
+   * @param {Object} displayName
+   * @param {Object} parent
```

Comment:
Where does the user get this object? Is this just the convo's activity object?

Reply:
correct. i'll elaborate in the comment.